requirements.txt

```diff
@@ -3,7 +3,10 @@ dipy
 # https://github.com/tensorflow/tensorflow/issues/44467
 h5py~=2.10.0
 Keras==2.3.1
-ivadomed
+# PyTorch's Linux distribution is very large due to its GPU support,
+# but we only need that for training models. Our users only need CPU.
+--extra-index-url https://download.pytorch.org/whl/cpu
```

Comment:
Up until now, I've been inferring the behavior of `--extra-index-url` based on what I've seen in the demos.

> Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cpu

But, I'm realizing I don't actually understand *why* the `--extra-index-url` is prioritized over [`--index-url`](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-0) (i.e. https://pypi.org/simple)?

In other words, why exactly is it that the CPU version of `torch` get installed, rather than the GPU version?

Reply:
Hmm. Reading https://github.com/pypa/pip/issues/8606, there's a deep discussion there, with the tl;dr being that pip treats the index URLs [with equal priority](https://github.com/pypa/pip/issues/8606#issuecomment-665554122). But, that doesn't exactly answer my question...

The closest thing I've come to a description of what we're trying to do is: https://github.com/pypa/pip/issues/8606#issuecomment-965659563

> As anecdotal usage feedback, I see requirements.txt files with `--extra-index-url` used as a prefix to specific packages as if it were a per-requirement option; of course, this isn't actually how `--extra-index-url` works, but the invocation succeeds and generally does what the user expects (apart from edge cases where everything explodes in exciting nasal-daemon sorts of ways).

Reply:
But again, I don't really get how pypi decides, when all that `ivadomed` specifies is `torch~=1.8.1`... 

* PyPI: [torch-1.8.1-cp37-cp37m-manylinux1_x86_64.whl](https://files.pythonhosted.org/packages/56/74/6fc9dee50f7c93d6b7d9644554bdc9692f3023fa5d1de779666e6bf8ae76/torch-1.8.1-cp37-cp37m-manylinux1_x86_64.whl#sha256=95b7bbbacc3f28fe438f418392ceeae146a01adc03b29d44917d55214ac234c9)
* torch/cpu: [torch-1.8.1+cpu-cp37-cp37m-linux_x86_64.whl](https://download.pytorch.org/whl/cpu/torch-1.8.1%2Bcpu-cp37-cp37m-linux_x86_64.whl)

Yet, the [latter gets installed](https://github.com/spinalcordtoolbox/spinalcordtoolbox/runs/6717573228?check_suite_focus=true#step:3:454):

```
Collecting torch~=1.8.1
  Downloading https://download.pytorch.org/whl/cpu/torch-1.8.1%2Bcpu-cp37-cp37m-linux_x86_64.whl (169.1 MB)
```

Reply:
From https://stackoverflow.com/a/67358322:

> Pip does not really prioritize one index over the other in theory. In practice, because of a coincidence in the way things are implemented in code, it might be that one is always checked first, but it is not a behavior you should rely on.

Reply:
Is it because we're not explicitly specifying `--index-url` in `requirements.txt`? Hmm...

Reply:
That''''''s a really good point. I read further down the thread and found some answers, from [ComputeCanada of all people](https://github.com/ComputeCanada/software-stack/blob/main/pip-which-version.md#build-tags-and-local-versions):

> If such a tag is found, it is used as a tie breaker between two otherwise identical packages.

So the existence of `+cpu` is what makes that wheel win.

Reply:
ComputeCanada is in the process of adopting torch's solution:

> As of July 2021, we have yet to repackage all of our existing python wheels to add local version specifiers, since this will be a long endeavour, but the newly compiled ones automatically include it. We are therefore still relying on a pip<21 constraint, but we should be able to lift that constraint in the future.

Of course the better solution would be if like, torch was split up in such a way that you could just say `torch[cpu]` or `torch[gpu]` directly.

Reply:
To get even closer to the source on this, the ComputeCanada docs say:

> These options were pointed to us in https://github.com/pypa/pip/issues/10156, by [Pradyun Gedam](https://github.com/pradyunsg).

https://github.com/pypa/pip/issues/10156#issuecomment-878571384

> Another option is wheel build tags -- which might be an even better answer. If a wheel has a build tag, that wheel is preferred over a wheel that doesn't.

Reply:
Aha, even closer to the source, https://github.com/pypa/pip/issues/10156#issuecomment-878606128:

> None the less, I think the local version will be prefered over build tags. The order of preference is specified at: https://github.com/pypa/pip/blob/main/src/pip/_internal/index/package_finder.py#L530

Reply:
https://peps.python.org/pep-0440/#summary-of-permitted-suffixes-and-relative-ordering is probably the closest thing to   the final word on priority.

Reply:
Sorry to bump an old thread but I've noticed the way CC's solution falls down. I mean, they're aware and @kanishk16 just quoted it:

>  Note that if one does not require an explicit version when installing the package, and if a more recent version of the package is uploaded on PyPI, pip will still prefer the more recent version.

But to be really clear, this means currently, asking for the cu101 build of torch instead falls over to the PyPI build:

```
$ pip install torch --extra-index-url https://download.pytorch.org/whl/cu101
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cu101
Collecting torch
  Downloading torch-1.11.0-cp310-cp310-manylinux1_x86_64.whl (750.6 MB)
     |█                               | 25.2 MB 10.7 MB/s eta 0:01:08^C
```


Though asking for the [`+cu102`](https://download.pytorch.org/whl/cu102/torch) build and above currently works as expected:

```
$ pip install torch --extra-index-url https://download.pytorch.org/whl/cu102
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cu102
Collecting torch
  Downloading https://download.pytorch.org/whl/cu102/torch-1.11.0%2Bcu102-cp310-cp310-linux_x86_64.whl (750.6 MB)
     |▊                               | 17.1 MB 928 kB/s eta 0:13:10^C
```

But at some point there will be a torch v1.12 and a torch v1.13 on PyPI that aren't not mirrored in https://download.pytorch.org/whl/cu102/torch, and that that point there will be a problem.

**But** this is safe for us because:

2. `ivadomed` currently [*is* pinning `torch`'s version](https://github.com/ivadomed/ivadomed/blob/f063485b083fc5503e1a1ee4ae0d7e947131603f/requirements.txt#L24) to `~=1.8.1`.

   It's possible with `~=` that they could upload a 1.8.3 to PyPI which would then take precedence over https://download.pytorch.org/whl/cpu/torch, but I think it's unlikely for them to do that. I think it's less likely they'll decide to skip uploading to the side-repos for a patch version than for a minor or major version. 
1. unlike the CUDA builds, torch is publishing a `+cpu` for every single version they put out, so we can trust that `+cpu` will break the tie in favor of the CPU build.

So leaning on local version tags is fragile, but should be okay for us.